activerecord/lib/active_record/inheritance.rb

```diff
@@ -198,12 +198,12 @@ def type_condition(table = arel_table)
       # If this is a StrongParameters hash, and access to inheritance_column is not permitted,
       # this will ignore the inheritance column and return nil
       def subclass_from_attributes?(attrs)
-        attribute_names.include?(inheritance_column) && (attrs.is_a?(Hash) || attrs.respond_to?(:permitted?))
+        attribute_names.include?(inheritance_column) && (attrs.is_a?(Hash) || attrs.respond_to?(:permitted?) || columns_hash[inheritance_column].default)
       end
 
       def subclass_from_attributes(attrs)
         attrs = attrs.to_h if attrs.respond_to?(:permitted?)
-        subclass_name = attrs.with_indifferent_access[inheritance_column]
+        subclass_name = (attrs || {}).with_indifferent_access[inheritance_column] || columns_hash[inheritance_column].default
```

Comment:
No, actually in my previous commit I was initializing `attrs ||= {}` so that's why I called `with_indifferent_access` directly, which is equivalent to current implementation.